numpy/core/src/multiarray/refcount.c.src

```diff
@@ -234,60 +180,87 @@ PyArray_XDECREF(PyArrayObject *mp)
         data = (PyObject **)PyArray_DATA(mp);
         n = PyArray_SIZE(mp);
         if (PyArray_ISALIGNED(mp)) {
-            for (i = 0; i < n; i++, data++) Py_XDECREF(*data);
+            for (i = 0; i < n; i++, data++) {
+                HANDLE_ITEM_AT(data);
+            }
         }
         else {
-            for (i = 0; i < n; i++, data++) {
-                NPY_COPY_PYOBJECT_PTR(&temp, data);
-                Py_XDECREF(temp);
+            for( i = 0; i < n; i++, data++) {
+                HANDLE_ITEM_AT_UNALIGNED(data);
             }
         }
     }
     else { /* handles misaligned data too */
         PyArray_RawIterBaseInit(&it, mp);
-        while(it.index < it.size) {
-            NPY_COPY_PYOBJECT_PTR(&temp, it.dataptr);
-            Py_XDECREF(temp);
+        while (it.index < it.size) {
+            data = (PyObject **)it.dataptr;
+            HANDLE_ITEM_AT_UNALIGNED(data);
             PyArray_ITER_NEXT(&it);
```

Comment:
I don't think this is ever hit

Reply:
Probably not, it is pretty rare to create unaligned object arrays. Same goes for object arrays actually holding NULL instead of None. Although, I think the first one at least may be simple enough with a structured dtype, which would indeed be a good test.

Reply:
nvm, made a test for this, I think.